torch/_dynamo/variables/dicts.py

```diff
@@ -1,43 +1,161 @@
 import collections
 import dataclasses
+import enum
 import functools
 import inspect
 import sys
 from typing import Any, Dict, List, Optional
 
-import torch
-import torch.fx
+from torch._subclasses.fake_tensor import is_fake
 
 from .. import variables
 from ..bytecode_transformation import create_call_function, create_instruction
 from ..eval_frame import skip_code
 
 from ..exc import unimplemented
 from ..guards import GuardBuilder, install_guard, make_dupe_guard
-from ..source import AttrSource, GetItemSource, GlobalWeakRefSource
-from ..utils import global_key_name, istensor, istype, iter_contains
+from ..source import AttrSource, GetItemSource
+from ..utils import istype, iter_contains, specialize_symnode
 from .base import MutableLocal, VariableTracker
 from .constant import ConstantVariable
-from .tensor import TensorVariable
+
+
+# Note: [Adding a new supported class the keys of ConstDictVarialble]
+# You'll need to add it to:
+# - `is_hashable_python_var` in this file
+# - `is_hashable` in this file
+# - `const_repr` in util.py, and perhaps modify DICT_KEYS in guards.py
+
+
+def is_hashable_python_var(x):
+    from torch import Tensor
+
+    # Note: Keep me in sync with is_hashable!
+    # Even better, we should have a map of functions connecting the two
+    from ..trace_rules import is_builtin_callable, is_numpy
+
+    return (
+        ConstantVariable.is_literal(x)
+        or isinstance(x, (Tensor, enum.Enum, type))
+        or is_builtin_callable(x)
+        or (isinstance(x, tuple) and all(is_hashable_python_var(e) for e in x))
+        or is_numpy(x)
+    )
+
+
+def is_hashable(x):
+    # Keep me in sync with is_hashable_python_var!
+    # Even better, we should have a map of functions connecting the two
+    if isinstance(x, variables.TensorVariable):
+        # Tensors are hashable if they have an example_value (a fake tensor)
+        # Most VT's should have one.
+        # It'd be nice if at some point we could assert that they all have one
+        return x.as_proxy().node.meta.get("example_value") is not None
+    elif isinstance(x, variables.TupleVariable):
+        return all(is_hashable(e) for e in x.items)
+    else:
+        return isinstance(
+            x,
+            (
+                variables.BuiltinVariable,
+                variables.SymNodeVariable,
+                variables.ConstantVariable,
+                variables.EnumVariable,
+                variables.user_defined.UserDefinedClassVariable,
```

Comment:
@voznesenskym this was not in the original PR. I think this will choke if we try to run a guard on it and execute this guard. Could you add a test in a separate PR?

I'm not sure about the SkipFiles variable either, we should add a test as well